workflows/rnaseq/downstream/functional-enrichment.Rmd

```diff
@@ -29,10 +29,50 @@ id.convert <- TRUE
 # Convert gene IDs to those useable to clusterProfiler.
 # Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#bitrgenes
 # NOTE: Gene IDs are provided in what format?----------------------------------
-from.type <- "FLYBASE"
+from.type <- 'ENSEMBL'
+annotation_genus_species <- 'Mus musculus'
+annotation_key_override <- NA
+hub.cache <- '../../../include/AnnotationHubCache'
+orgdb <- get.orgdb(
+    annotation_genus_species,
+    cache=hub.cache,
+    annotation_key_override=annotation_key_override
+)
+# In place of AnnotationDbi::select, Use a wrapper function to run this command in order to catch the unmatching keys error
+# The func parameter corresponds to AnnotationDbi::select
+# The pattern parameter is used to identify the error raised when you have genes, but these genes do not encode proteins
+# This may happen as functional enrichment is only defined for proteins as of now. 
+# This is the common form of the error message: "Error in .testForValidKeys(x, keys, keytype, fks) : 
+#"None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments."
+# You do not need to pass this entire thing to the wrapper, just enough that it can be identified with a grep search. 
+# When the exception is caught, assign the contrast a null value and interpret that value downstream in the Rmd.
+keyhandler <- function(func, pattern,  ...) {
+
+    wrapped_select <- function(...) {
+        tryCatch(
+            {
+             func(...)
+            },
+            error = function(e){
+                msg <- conditionMessage(e)
+                if (grepl(pattern, msg)){
+                    message("Caught a valid key exception in AnnotationDbi::select. Confirm that, for contrasts with no functional enrichment, the genes do not encode proteins")
+
+                    return(NULL)
+                    } else {
+                    stop(msg)
+                }
+            }
+        )
+    }
+    wrapped_select
+}
+select.keyhandler <- keyhandler(AnnotationDbi::select, 'None of the keys')
 
 deg.list <- nested.lapply(sel.list, rownames)
-genes.df <- nested.lapply(deg.list, function (x) AnnotationDbi::select(orgdb, keys=x, columns=types, keytype=from.type))
+genes.df <- nested.lapply(deg.list, function (x) select.keyhandler(orgdb, keys=x, columns=types, keytype=from.type))
```

Comment:
if the function is moved to the lcdbwf package, then this would become:

```suggestion
genes.df <- nested.lapply(deg.list, function (x) lcdbwf::select.keyhandler(orgdb, keys=x, columns=types, keytype=from.type))
```